File: src/main/java/com/svantle/score/commands/FlySpeedCommand.java
package com.svantle.score.commands;

import com.svantle.score.utils.MessagesHandler;

import net.kyori.adventure.text.format.NamedTextColor;
import org.bukkit.command.Command;
import org.bukkit.command.CommandExecutor;
import org.bukkit.command.CommandSender;
import org.bukkit.entity.Player;
import net.kyori.adventure.text.Component;

public class FlySpeedCommand implements CommandExecutor {

    private final MessagesHandler messagesHandler;

    public FlySpeedCommand(MessagesHandler messagesHandler) {
        this.messagesHandler = messagesHandler;
    }

    @Override
    public boolean onCommand(CommandSender sender, Command cmd, String label, String[] args) {
        if (!(sender instanceof Player player)) {
            sender.sendMessage(messagesHandler.prefixed(messagesHandler.get("only_players")));
            return true;
        }

        if (!player.hasPermission("score.flyspeed")) {
            player.sendMessage(messagesHandler.prefixed(messagesHandler.get("no_permission")));
            return true;
        }

        if (args.length != 1) {
            player.sendMessage(messagesHandler.prefixed(messagesHandler.get("flyspeed_usage")));
            return true;
        }

        setPlayerFlySpeed(player, args[0]);

        return true;
    }

    private void setPlayerFlySpeed(Player player, String input) {
        try {
            int speed = Integer.parseInt(input);

            if (speed < 1 || speed > 10) {
                player.sendMessage(messagesHandler.prefixed(messagesHandler.get("flyspeed_invalid")));
                return;
            }

            float actualSpeed = speed / 10.0f;
            player.setFlySpeed(actualSpeed);
            player.sendMessage(messagesHandler.prefixed(messagesHandler.get("flyspeed_updated").append(Component.text(String.valueOf(speed), NamedTextColor.GOLD))));

        } catch (NumberFormatException e) {
            player.sendMessage(messagesHandler.prefixed(messagesHandler.get("flyspeed_invalid")));
        }
    }
}

File: src/main/kotlin/com/svantle/score/commands/FlySpeedCommand.kt
package com.svantle.score.commands

import com.svantle.score.SCore
import com.svantle.score.constants.*
import com.svantle.score.utils.MessageHelper
import com.svantle.score.utils.sendTo
import org.bukkit.Bukkit
import org.bukkit.command.Command
import org.bukkit.command.CommandExecutor
import org.bukkit.command.CommandSender
import org.bukkit.entity.Player

class FlySpeedCommand(core: SCore) : CommandExecutor {

	private val minSpeed = 1
	private val maxSpeed = 10

	init {
		core.getCommand("flyspeed")!!.setExecutor(this)
	}

	override fun onCommand(sender: CommandSender, cmd: Command, label: String, args: Array<String>?): Boolean {
		if(args.isNullOrEmpty()) {
			MessageHelper.phrase(MSG_COMMAND_USAGE, "/flyspeed <speed> [player]").sendTo(sender)
		} else if (args.size == 1) {
			handleFlySpeedSelf(sender, args[0])
		} else {
			handleFlySpeedOthers(sender, args[0], args[1])
		}

		return true
	}

	private fun validateSpeed(speed: String): Boolean {
		return speed.toIntOrNull() != null && speed.toInt() in minSpeed..maxSpeed
	}

	private fun handleFlySpeedSelf(sender: CommandSender, speed: String) {
		if(sender !is Player) {
			MessageHelper.phrase(MSG_ONLY_PLAYERS).sendTo(sender)
			return
		}

		if(sender.hasPermission(FLY_SPEED_COMMAND)) {
			if (validateSpeed(speed)) {
				val newSpeed = speed.toFloat() / 10.0f

				sender.flySpeed = newSpeed

				MessageHelper.phrase(MSG_FLY_SPEED_UPDATED, speed).sendTo(sender)
			} else {
				MessageHelper.phrase(MSG_INVALID_NUMBER, speed, minSpeed, maxSpeed).sendTo(sender)
			}
		} else {
			MessageHelper.phrase(MSG_NO_PERMISSION).sendTo(sender)
		}
	}

	private fun handleFlySpeedOthers(sender: CommandSender, speed: String, targetName: String) {
		val target = Bukkit.getPlayer(targetName)

		if (target == null) {
			MessageHelper.phrase(MSG_PLAYER_NOT_FOUND, targetName).sendTo(sender)
			return
		}

		if (sender.hasPermission(FLY_SPEED_COMMAND_OTHERS) || sender !is Player) {
			if (validateSpeed(speed)) {
				val newSpeed = speed.toFloat() / 10.0f

				target.flySpeed = newSpeed

				MessageHelper.phrase(MSG_FLY_SPEED_UPDATED, speed).sendTo(sender)
				MessageHelper.phrase(MSG_FLY_SPEED_UPDATED_OTHERS, target.name, speed).sendTo(target)
			} else {
				MessageHelper.phrase(MSG_INVALID_NUMBER, speed, minSpeed, maxSpeed).sendTo(sender)
			}
		} else {
			MessageHelper.phrase(MSG_NO_PERMISSION).sendTo(sender)
		}
	}
}
